### PR TITLE
Fixes issues 22185, 22673

### PR DESCRIPTION
**Pull request by maxhaton**
Currently uses an assert, open to suggestions to do something else (or nothing)

### CODE REVIEW

File: std/array.d [hunk 1]
```diff
@@ -126,12 +126,33 @@ if (isIterable!Range && !isAutodecodableString!Range && !isInfinite!Range)
...
+                emplaceRef!E(result[cnt], e);
+                ++cnt;
+            }
+        } catch (Exception e)
+        {
+            //https://issues.dlang.org/show_bug.cgi?id=22185
+            //Make any uninitialized elements safely destructible.
+            foreach(ref elem; result[cnt..$])
+            {
+                emplaceRef!E(elem, E.init);
```
Comment (kinke) [inline std/array.d:143]: `emplaceRef!E(elem);` uses cheap `emplaceInitializer`: https://github.com/dlang/druntime/blob/9d99b1ccd90d1e66c31e27cc71b8a266612aa14a/src/core/internal/lifetime.d#L23

Comment (maxhaton) [reply std/array.d:143]: Ok lets use that instead

**Comment by maxhaton** [comment]
@atilaneves Thoughts on the assertion?

File: std/array.d [hunk 2]
```diff
@@ -439,6 +461,90 @@ if (isAutodecodableString!String)
...
+            throw new Exception("This throws");
+        }
+        ~this()
+        {
+            /*
+                Any time this destructor runs, it should be running on "valid"
+                data. This is is mimicked by having a .init other than 0 (the value the memory
+                practically will be from the GC).
+            */
+            if (x != 420) {
```
Comment (thewilsonator) [inline std/array.d:510]:
```
Enforce Allman style
--
  | grep -nrE '(if\|for\|foreach\|foreach_reverse\|while\|unittest\|switch\|else\|version) .*{ $(find etc std -name '*.d'); test $? -eq 1
  | std/array.d:510:            if (x != 420) {
```

**Review by thewilsonator — APPROVED** [review]

File: std/array.d [hunk 3]
```diff
@@ -126,12 +126,34 @@ if (isIterable!Range && !isAutodecodableString!Range && !isInfinite!Range)
...
         }
+        /*
+            https://issues.dlang.org/show_bug.cgi?id=22673
+
+            We preallocated an array, we should ensure that enough range elements
+            were gathered such that every slot in the array is filled. If not, the GC
+            will collect the allocated array, leading to the `length - cnt` left over elements
+            being collected too - despite their contents having no guarantee of destructibility.
+         */
+        assert(length == cnt, "Range .length property was not equal to number of elements yielded by the range before becoming empty");
```
Comment (thewilsonator) [inline std/array.d:156]:
```
std/array.d(156:31)[warn]: Line is longer than 120 characters
```

**Comment by atilaneves** [comment]
> @atilaneves Thoughts on the assertion?

Works for me.

**Comment by RazvanN7** [comment]
@maxhaton Please take care of the style warning so that we can get this in.

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
@kinke is it possible that the circleci failure is related to your recent template emission strategy fix? I just rebased this PR and it is now failing:

```
std_uni_package.o: In function `_D3std6format__TQkVAyaa68_737461636b5b6b202d20325d2e6c656e677468202573203e20737461636b5b6b202d20315d2e6c656e677468202573202b20737461636b5b6b5d2e6c656e677468202573TmTmTmZQGhFNaNfmmmZQGg':
/home/circleci/phobos/std/format/package.d:1384: undefined reference to `_D3std6format5write__T14formattedWriteTSQBm5array__T8AppenderTAyaZQoTaTmTmTmZQCdFNaNfKQBvMxAammmZk'
```

**Comment by kinke** [comment]
@RazvanN7: Yeah, that's probably it. I'm not sure what the betterC tests extractor does exactly, but it looks like it kills the other non-`@betterC` unittests and then compiles each module separately with `-unittest -betterC`. One of the betterC problems is that it doesn't affect template emission at all - while druntime and Phobos aren't linked, template culling still behaves as usual, not emitting any template instances that are guaranteed to have been emitted into druntime/Phobos too. So old `-unittest` (amounting to something people would expected from `-allinst`) seems to have enabled such tests, while it's no guarantee that it then actually works for a user when compiling without `-unittest`. New `-unittest` now shows that problem. And as `-allinst` still doesn't really emit all required template instances, it probably won't be enough either. So unfortunately, I don't know how to easily fix this.

**Comment by kinke** [comment]
Ah wait, the first failure doesn't seem to be an extraction of `@betterC` tests, but public unittests. According to the Makefile comments, those tests are extracted to a separate module, compiled with `-unittest -version=StdUnittest` and linked against `libphobos2.a` (which has NOT been compiled with `-version=StdUnittest`). So I guess the problem is exactly that one, a template being culled because it should have been emitted into druntime/Phobos *if* compiled with `StdUnittest`.

**Comment by MoonlightSentinel** [comment]
> @RazvanN7: Yeah, that's probably it. I'm not sure what the betterC tests extractor does exactly, but it looks like it kills the other non-`@betterC` unittests and then compiles each module separately with `-unittest -betterC`.

`publictests` uses libdparse to move documented unittests into a distinct module to emulate users copy + pasting the examples. This is done for "normal" and Better-C compatible tests (tagged with `@betterC`) - the latter are compiled with `-betterC`.


> Ah wait, the first failure doesn't seem to be an extraction of @betterC tests, but public unittests. According to the Makefile comments, those tests are extracted to a separate module, compiled with -unittest -version=StdUnittest and linked against libphobos2.a (which has NOT been compiled with -version=StdUnittest). So I guess the problem is exactly that one, a template being culled because it should have been emitted into druntime/Phobos if compiled with StdUnittest.

Seems like the test extractor should omit `-version=StdUnittest` (and maybe use explicit functions instead of unittests)

**Comment by RazvanN7** [comment]
>Ah wait, the first failure doesn't seem to be an extraction of @betterC tests, but public unittests. According to the Makefile comments, those tests are extracted to a separate module, compiled with -unittest -version=StdUnittest and linked against libphobos2.a (which has NOT been compiled with -version=StdUnittest). So I guess the problem is exactly that one, a template being culled because it should have been emitted into druntime/Phobos if compiled with StdUnittest.

But the link failure appears to be caused by a line inside std.format.format, not inside the unittest. The missing symbol is `formattedWrite`  inside [std.format.format](https://github.com/dlang/phobos/blob/master/std/format/package.d#L1384). I would expect that this instantiation have nothing to do with `version(StdUnittest)`. I can't see anything related to `version(StdUnittest)` in `std.format.write`.

> Seems like the test extractor should omit -version=StdUnittest (and maybe use explicit functions instead of unittests)

To me it sounds like the test extractor should should also copy-paste the `version(StdUnittest)` blocks into the test module. However, at this point I am not convinced that this is the issue.

**Comment by RazvanN7** [comment]
> Right - I hope they still work and don't depend on anything hidden behind `version(StdUnittest)`.

`version(StdUnittest)` is used to define helper functions for unittests. So to run the unittests these are required, therefore the test extractor needs to somehow make sure that the code is available (probably by also pasting the body of the version block).

**Comment by MoonlightSentinel** [comment]
> `version(StdUnittest)` is used to define helper functions for unittests. So to run the unittests these are required, therefore the test extractor needs to somehow make sure that the code is available (probably by also pasting the body of the version block).

No, such code should not appear in a documented unittest because they are included as examples on dlang.org. User won't have access to those hidden symbols.

**Comment by RazvanN7** [comment]
Oh I see. So the test extractor picks up only the documented unittests. I missed the "documented" part. Yeah, it sucks that we are using inexistent functions in documented unittests (Ctrl-f `ceqrel` on https://dlang.org/phobos/std_complex.html )

**Comment by RazvanN7** [comment]
> But the link failure appears to be caused by a line inside std.format.format, not inside the unittest. The missing symbol is formattedWrite inside std.format.format. I would expect that this instantiation have nothing to do with version(StdUnittest). I can't see anything related to version(StdUnittest) in std.format.write.


I was wrong. The problem is indeed coming from using `-version=StdUnittest` for the extracted tests. I tested this and eliminating it makes the linker error go away, however, there are a few cases where documented unittests depend on symbols hidden behind `version(StdUnittest)`. I will make PR to fix this breakage.

**Pull request merged**